File: pyversion/40. combination sum II.py
def combiSum2(nums, target):
    nums.sort()
    res = []
    findCombi(nums, 0, [], res, target)
    return res


def findCombi(nums, start, path, res, target):

    if not target:
        res.append(path)
        return
    for i in range(start, len(nums)):
        if i > start and nums[i] == nums[i - 1]:
            continue
        if nums[i] > target:
            break
        findCombi(
            nums,
            i + 1,
            path + [nums[i]],
            res,
            target - nums[i],
        )


if __name__ == '__main__':
    print(combiSum2([10, 1, 2, 7, 6, 1, 5], 8))

File: cppversion/40. combination sum II.cpp
#include <algorithm>
#include <iostream>
#include <vector>

void findCombi(std::vector<int> &nums, std::vector<int> &candidates, int target, int begin_index, std::vector<std::vector<int>> &res);

std::vector<std::vector<int>>
combiSum2(std::vector<int> &nums, int target)
{

    std::sort(nums.begin(), nums.end());
    std::vector<std::vector<int>> res;
    std::vector<int> candidates;
    findCombi(nums, candidates, target, 0, res);
    return res;
}

void findCombi(std::vector<int> &nums, std::vector<int> &candidates, int target, int begin_index, std::vector<std::vector<int>> &res)
{
    if (target == 0)
    {
        res.push_back(candidates);
        return;
    }

    else
    {
        for (int i = begin_index; i < nums.size(); i++) // iterative
        {
            if (nums[i] > target)
                return;
            if ((nums[i] == nums[i - 1]) and (i > begin_index))
                continue; // check duplicate combinations
            candidates.push_back(nums[i]);
            findCombi(nums, candidates, target - nums[i], i + 1, res); // recursive
            candidates.pop_back();
        }
    }
}

int main()
{
    std::vector<int> nums{10, 1, 2, 7, 6, 1, 5};
    int target = 9;
    std::vector<std::vector<int>> res;
    res = combiSum2(nums, target);
    for (auto values : res)
    {
        for (auto v : values)
        {
            std::cout << v << " ";
        }
        std::cout << std::endl;
    }
    std::cin.get();
}
